DiskStorage Class

aluamorim opened the pull request:
I have added a DiskStorage class to the 'storage.py' source file. The class behaves just like the NumpyStorage class but saving/loading checkpoints to/from a .dat binary file created in the constructor. The '.dat' file is created in the working directory when the DiskStorage class is instantiated and removed when it is destroyed. All .dat files are named with a timestamp using the following template string:

"CKP_{}.dat".format(datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))"

Checkpoints are stored sequentially according to their 'key' number.

navjotk commented:
Thank you for the contribution. This looks promising. A few requests:
1. Can you add a couple of tests to show this working please?
2. The commit messages could be more descriptive
3. There should probably be a way to provide the path to the directory where the checkpoints are to be stored
4. There are many ways to read and write numpy data to/from a file. We should compare them to decide which is faster/slower. Have you done some experiments around this already?

aluamorim commented:
Hi, regarding your comments:

1. Can you add a couple of tests to show this working please?
I have added two test units to the test/ folder, one for the MemoryRevolver and another for the DiskRevolver.

2. The commit messages could be more descriptive
I have added more detail to the new commit message.

3. There should probably be a way to provide the path to the directory where the checkpoints are to be stored
Now the DiskStorage and DiskRevolver classes also receive a filename as a parameter, so the user can decide whether to use the default .dat file or provide another.

4. There are many ways to read and write numpy data to/from a file. We should compare them to decide which is faster/slower. Have you done some experiments around this already?
I tested two different other ways, but using the Numpy save and load functions was faster. I'll look it up to see if it is the same for very large arrays.

navjotk commented:
Waiting for Travis CI to be enabled on this repo again to proceed.

navjotk commented:
> I tested two different other ways, but using the Numpy save and load functions was faster. I'll look it up to see if it is the same for very large arrays.
Can you post the methods you tried and your results please?

> I have added more detail to the new commit message.
What I would like to see is that when I go to the "commits" tab on this page, I should be able to understand what each commit is doing without having to click through. None of the commits help me do that right now. Can you please use `git rebase -i` to fix the commit messages to achieve this?

Other things:
- Have you tried running `flake8 .` on this PR? I see some issues (which is how I realised CI wasn't running).
- There are binary files being included in this PR. Can you remove them please?
- Can we write tests that don't import Devito? Devito includes pyrevolve as a dependency. If we can't test pyrevolve without devito, this makes it a cyclic dependency.

navjotk commented:
@aluamorim can you please `git merge master` in your branch so we can have the CI that @ofmla created, working on this PR? Thanks!

navjotk commented on pyrevolve/pyrevolve.py:
```diff
@@ -71,32 +68,26 @@ def __init__(self, checkpoint, fwd_operator, rev_operator,
-
-        self.scheduler = Revolve(n_checkpoints, n_timesteps)
+        self.scheduler = Revolve(self.n_checkpoints, n_timesteps)
+        self.storage = storage
```
Seems to have left a weird corner case where, if the user did not provide a storage, it would be initialised to None? That is not a useful configuration. It would be better to have useful defaults.

aluamorim replied:
The idea was to make the BaseRevolver class abstract so that the user can only instantiate either MemoryRevolver or DiskRevolver, but I forgot to make it explicit. In the next commit, all make the BaseRevolver class abstract.

navjotk commented on pyrevolve/pyrevolve.py:
```diff
@@ -163,3 +153,59 @@ def save_checkpoint(self):
+    """
+    def __init__(self, checkpoint, fwd_operator, rev_operator,
+                 n_checkpoints=None, n_timesteps=None, timings=None,
+                 filename=None, remove_storagefile=True):
```
Wouldn't it be better to take a directory instead of a filename? Then we can try to generate unique filenames instead of leaving that to the user?
Also, is the remove_storagefile flag necessary? Why not just remove it every time?

aluamorim replied:
Agreed.

navjotk commented on pyrevolve/storage.py:
```diff
@@ -1,10 +1,95 @@
+        self.profiler = profiler
+        if filename is None:
+            self.filename = "CKP_{}.dat".format(
+                datetime.datetime.now().strftime("%Y%m%d-%H%M%S"))
```
We should be doing something better than this to generate filenames. Think shared filesystem and MPI. We wouldn't want multiple ranks clashing? Minimum would be to include PID in the filename. Ideally we want to generate names in a better way.

aluamorim replied:
You're. I have added the thread Id to the filename.

navjotk replied:
It should be the process ID. Threads are sort of irrelevant to MPI.

navjotk commented on pyrevolve/storage.py:
```diff
@@ -1,10 +1,95 @@
+
+    def setW(self):
+        ''' Sets WRITE stream as default '''
+        self.default_storage_stream = self.storage_w
```
What's going on here?

aluamorim replied:
"Load" and "save" methods access "read" and "write" file streams respectively, but both use the same implementation of __getitem__(). I had to switch streams explicitly to avoid messing with the file pointer (seek), because the "appy_reverse"  method in alternates between load and save calls.

navjotk replied:
I was actually wondering about this. Why do we choose to store everything in one file? Why not one file per checkpoint, for example?

navjotk commented on tests/.gitignore:
```diff
@@ -0,0 +1 @@
+__pycache__
```
Please uncommit this file and the pyc file

aluamorim replied:
OK

navjotk commented:
Also, there is the question of whether all checkpoints need to be stored in the same file or different ones. Is there a reason to make the choice that we did?

navjotk commented on pyrevolve/storage.py:
```diff
@@ -8,54 +8,65 @@
-        self.filename = self.filedir + "CKP_{}_{}.dat".format(
+        self.filename = self.filedir + "CKP_D{}_T{}.dat".format(
             datetime.datetime.now().strftime("%Y%m%d-%H%M%S"),
             threading.get_ident())
```
We need `os.getpid()` here instead

navjotk commented on tests/utils.py:
```diff
@@ -77,3 +77,56 @@ def dtype(self):
     def size(self):
         print("bytes", self.field.nbytes)
```
Delete

navjotk commented on tests/utils.py:
```diff
@@ -77,3 +77,56 @@ def dtype(self):
+            self.u[:] = self.u[:] + self.direction * abs(t_start - t_end)
+        else:
+            self.v[:] = (self.u[:]*(-1) + 1)
+            # print(self.v, flush=True)
```
Delete

navjotk commented on tests/test_diskckp.py:
```diff
@@ -0,0 +1,105 @@
+                       filedir="./", singlefile=False)
+    rev.apply_forward()
+    assert(b.counter == 0)
+    print(df)
```
Please remove all prints

navjotk commented on tests/test_diskckp.py:
```diff
@@ -0,0 +1,105 @@
+
+@pytest.mark.parametrize("nt, ncp", [(10, 2), (10, 4), (10, 6), (10, 8),
+                                     (10, 9), (10, 10), (10, 11), (10, 12)])
+def test_forward_nt_multifile(nt, ncp):
```
These two tests can be combined by parametrising on single file.

navjotk reviewed (changes requested):
There are leftover prints all over the place. Plus I don't like the fact that we have copy pasted the same test 4 times to change one parameter. We should be using pytest's test parametrisation to do this in a single line. Also some binary files have made their way in again. Use .gitignore

navjotk commented:
Thanks for the cooperation so far. Two last things please:
- There is still a binary pyc file committed ( tests/__pycache__/test_pyrevolve.cpython-27-PYTEST.pyc). Can you make sure to remove it?
- Can we also parametrise on the type of storage so we can combine `test_diskckp.py` and `test_memckp.py`?

navjotk commented:
I can still see a committed file called `tests/__pycache__/test_pyrevolve.cpython-27-PYTEST.pyc`. Am I missing something?

Can we rename the file from `test_diskckp.py` to something more appropriate?

Also, this PR now has 147 commits - which is about the same as the total number of commits in pyrevolve so far :) Can we please use `git rebase -i` ([tutorial](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec)) to squash all the commits into just one?

aluamorim commented:
> I can still see a committed file called `tests/__pycache__/test_pyrevolve.cpython-27-PYTEST.pyc`. Am I missing something?
That is strange, I removed this file from the recent commits and then I purged it from all git history. It doesn't appear for me anymore, not even in previous commits. I don't know what is going on.
 
> Can we rename the file from `test_diskckp.py` to something more appropriate?

Sure. Do you suggest any name? It tests both disk and memory storage classes, would it make sense if I moved all the tests on this file to test_storage.py?

> Also, this PR now has 147 commits - which is about the same as the total number of commits in pyrevolve so far :) Can we please use `git rebase -i` ([tutorial](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec)) to squash all the commits into just one?

I'll do it right away.
